ext/oboe_metal/src/frames.cc

```diff
@@ -0,0 +1,244 @@
+mutex cached_frames_mutex;
+
+void Frames::reserve_cached_frames() {
+    lock_guard<mutex> guard(cached_frames_mutex);
+    // unordered_maps grow automatically, but it starts at 1 and then
+    // doubles when it is full, so lets avoid the warmup
+    cached_frames.reserve(500);  // it will round to a prime number: 503
+}
+
+void Frames::clear_cached_frames() {
+    lock_guard<mutex> guard(cached_frames_mutex);
+        // unordered_maps grow automatically, but it starts at 1 and then
+        // doubles when it is full, so lets avoid the warmup
+    cached_frames.clear();
+}
+
+// this is a private function
+int Frames::cache_frame(VALUE frame) {
+    VALUE val;
+    FrameData data;
+
+    // only cache it if it does not exist
+    if (cached_frames.count(frame) == 0) {
+        val = rb_profile_frame_label(frame);  // returns method or block
+        if (RB_TYPE_P(val, T_STRING))
+            data.method = RSTRING_PTR(val);
+
+        if (data.method.rfind("block ", 0) == 0) {
+            // we don't need more info if it is a block
+            // we ignore block level info because they make things messy
+            lock_guard<mutex> guard(cached_frames_mutex);
+            cached_frames.insert({frame, data});
+            return 0;
+        }
+
+        val = rb_profile_frame_classpath(frame);  // returns class or nil
+        if (RB_TYPE_P(val, T_STRING)) data.klass = RSTRING_PTR(val);
+
+        val = rb_profile_frame_absolute_path(frame);  // returns file, use rb_profile_frame_path() if nil
+        if (!RB_TYPE_P(val, T_STRING)) val = rb_profile_frame_path(frame);
+        if (RB_TYPE_P(val, T_STRING)) data.file = RSTRING_PTR(val);
+
+        // Ruby 3 reports <cfunc>, but the linenumbers are bogus
+        // the default line number is 0
+        if (!data.file.compare("<cfunc>") == 0) {
+            val = rb_profile_frame_first_lineno(frame);  // returns line number
+            if (RB_TYPE_P(val, T_FIXNUM)) {
+                data.lineno = NUM2INT(val);
+            }
+        }
+        lock_guard<mutex> guard(cached_frames_mutex);
+        cached_frames.insert({frame, data});
+    }
+    return 0;
+}
+
+// all frames in frames_buffer must be in cached_frames
+// before calling this function
+// we are saving the check for better performance
+int Frames::collect_frame_data(VALUE *frames_buffer, int num, vector<FrameData> &frame_data) {
+    if (num == 1) {
+        if (frames_buffer[0] == PR_IN_GC) {
+            FrameData data;
+            data.method = "GARBAGE COLLECTION";
+            frame_data.push_back(data);
+            return 0;
+        } else if (frames_buffer[0] == PR_OTHER_THREAD) {
+            FrameData data;
+            data.method = "OTHER THREADS";
+            frame_data.push_back(data);
+            return 0;
+        }
+    }
+
+    for (int i = 0; i < num; i++) {
+        VALUE frame = frames_buffer[i];
+        frame_data.push_back(cached_frames[frame]);
```

Comment:
I don't think it is necessary for reading, because the value at an index will not change. 
I added the mutex when writing because I'm not sure if it is thread safe when the map has to grow.

Reply:
From https://stackoverflow.com/questions/44635678/c-one-writer-multi-reader-access-to-an-unordered-map-thread-safe:

> The rule for this is: If you have shared data accessed by more than one thread and at least one of them is a writer, you must have synchronization.